src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,159 @@
+    private Node<E> tail;
+    private int size;
 
-public class MyLinkedList<T> implements MyLinkedListInterface<T> {
     @Override
-    public void add(T value) {
+    public void add(E value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            ++size;
+            return;
+        }
+        Node<E> newNode = new Node<>(tail, value, null);
+        tail.next = newNode;
+        tail = newNode;
+        ++size;
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(E value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<E> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            ++size;
+            return;
+        }
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        Node<E> add = new Node<>(newNode.prev, value, newNode);
+        newNode.prev.next = add;
+        newNode.prev = add;
+        size++;
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public void addAll(List<E> list) {
+        for (E element : list) {
+            add(element);
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public E get(int index) {
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        return newNode.value;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public E set(E value, int index) {
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        E result = newNode.value;
+        newNode.value = value;
+        return result;
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public E remove(int index) {
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        E removeValue = newNode.value;
+        unlink(newNode);
+        return removeValue;
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(E object) {
+        Node<E> newNode = head;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(object, newNode.value)) {
+                unlink(newNode);
+                return true;
+            }
+            newNode = newNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        if (size > 0) {
+            return false;
+        }
+        return true;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Not correct index " + index);
+        }
+    }
+
+    private Node<E> findNode(int index) {
+        Node<E> newNode;
+        if (index < size) {
```

Comment:
чи є нам сенс ітерувати ліст якщо індекс більший за розмір самого ліста? 
давай додамо checkIndex на початок цього методу 
```suggestion
        if (index < size / 2) {
```